<a href="https://colab.research.google.com/github/BalkeesRekik/DS1002/blob/main/Copy_of_14_TextGeneration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Final homework assignment:  generating text from a corpus

Due Friday 3 May at 5 pm

We begin by loading all of Shakespeare's plays into a text file, and separating them into individual words.  This is made easy by the fact that someone already made a giant text file of these plays for us, and put it online.

[Note that by changing the url in the following code chunk, you can load any text file that lives online in this way . . . see below.]

Run the following code.  It should output a list of ten words, starting with 'From' and ending with 'rose'.

In [1]:
from urllib.request import urlopen            #package for reading info from the web
url='https://ocw.mit.edu/ans7870/6/6.006/s08/lecturenotes/files/t8.shakespeare.txt'   #web address
text = urlopen(url).read().decode('utf-8')    #open web address, read file, convert to text
shakespeare_words=text.split()[1651:]         #split text into words (not including preamble)
shakespeare_words[:10]                        #display the first 10 words

['From',
 'fairest',
 'creatures',
 'we',
 'desire',
 'increase,',
 'That',
 'thereby',
 "beauty's",
 'rose']

## 1. `nextwords` practice

Create a practice function called `createNextwordsPractice` which takes a list of words called `wordlist` and returns a dictionary called `nextwordsPractice`. The keys of `nextwordsPractice` should be all the words in `wordlist`, and the value for any given key should be a list of all the words that follow the key word somewhere in the original `wordlist`.



In [8]:
def createNextwordsPractice(wordlist):
  nextwordsPractice = {}
  for i in range(len(wordlist)-1):
    key = wordlist[i]
    next_word = wordlist[i+1]
    if key in nextwordsPractice:
      nextwordsPractice[key].append(next_word)
    else:
        nextwordsPractice[key] = [next_word]
  return nextwordsPractice

# Test
print(createNextwordsPractice(shakespeare_words[:10]  ))

{'From': ['fairest'], 'fairest': ['creatures'], 'creatures': ['we'], 'we': ['desire'], 'desire': ['increase,'], 'increase,': ['That'], 'That': ['thereby'], 'thereby': ["beauty's"], "beauty's": ['rose']}


Run the following code:

In [9]:
practice = createNextwordsPractice(shakespeare_words)
practice['From']

['fairest',
 'his',
 'sullen',
 'limits',
 'hands',
 'whence',
 'where',
 'me',
 'this',
 'thy',
 'hence',
 'thee,',
 'you',
 'heaven',
 'her',
 'simple',
 'whence',
 'below',
 'lowest',
 'courtly',
 'son',
 'the',
 'the',
 'Sicyon,',
 'Alexandria',
 'Silvius,',
 'mine',
 'the',
 'firm',
 'Egypt',
 'Antony',
 'thine',
 'that',
 'my',
 'me',
 'henceforth',
 'tyrant',
 'seventeen',
 'miserable',
 'whom',
 'my',
 'whence,',
 'me',
 'me',
 'slaves',
 'every',
 'rascals',
 'whom',
 'where',
 "th'",
 'Rome',
 'off',
 'these',
 'twelve',
 "th'",
 'whence?',
 'Rome.',
 'thy',
 'Gallia',
 'fairies',
 'off',
 "th'",
 'most',
 'thy',
 'youth',
 'every',
 'whose',
 'so',
 'one',
 'this',
 'the',
 'east',
 'action',
 'my',
 'this',
 'Leonati',
 'her',
 'chance',
 'south',
 'our',
 'whence',
 'the',
 'top',
 'this',
 'our',
 'that',
 'me,',
 'fashion',
 'the',
 'another',
 'Hamlet?',
 'whence',
 'whom',
 "Rumour's",
 'whence',
 'the',
 'a',
 'a',
 'enemies',
 'every',
 'this',
 'the',
 'open',
 'who

In [10]:
practice['rose']

['might',
 'looks',
 'is',
 'of',
 'Of',
 'may',
 'will',
 'of',
 'From',
 'and',
 'of',
 'both',
 'with',
 'from',
 'with',
 'with',
 'side.',
 'red,',
 'too.',
 'in',
 'a',
 'a',
 'I',
 'against',
 'that',
 'and',
 'from',
 'again,',
 'Lest',
 'against',
 'Than',
 "distill'd",
 'on',
 'up',
 'in',
 'wither.',
 'and',
 'By',
 'again?',
 'he']

In [11]:
practice['rose.']

['It']

## 2. `nextwords` for real:  new sentences.

Next, write a new function called `createNextwords` which takes a `wordlist` and returns a dictionary of `nextwords`, but now with a new additional feature.  
In addition to the keys from problem 1, add a special key to the dictionary, which will be the symbol '%'.  This key will signify the first word of a sentence. When scanning the wordlist to construct the `nextwords` dictionary, normally you would add `wordlist[i+1]` to the list `nextwords[wordlist[i]]`. However, if `wordlist[i]` ends with a period, then you should instead add `wordlist[i+1]` to the list nextwords['%']. Also add the very first word of `wordlist` to nextwords['%'].


You should be able to produce this output:
```
> shakespeare_nextwords=createNextwords(shakespeare_words)
> shakespeare_nextwords['Rose']
['at', 'or', 'or', 'and']

> shakespeare_nextwords['rose.']
[]

> len(shakespeare_nextwords['%'])
76543
```

In [12]:
def createNextwords(wordlist):
  nextwords = {}
  nextwords['%'] = [wordlist[0]]

  for i in range(len(wordlist) - 1):
      current_word = wordlist[i]
      next_word = wordlist[i + 1]

      if current_word.endswith('.'):
          if '%' in nextwords:
              nextwords['%'].append(next_word)
          else:
              nextwords['%'] = [next_word]
          nextwords[current_word] = []
      else:
          if current_word in nextwords:
              nextwords[current_word].append(next_word)
          else:
              nextwords[current_word] = [next_word]

  return nextwords

Run the following code:

In [13]:
shakespeare_nextwords = createNextwords(shakespeare_words)

print(shakespeare_nextwords['Rose'])
print(len(shakespeare_nextwords['%']))

['at', 'or', 'or', 'and']
76543


In [14]:
print(len(shakespeare_nextwords['rose']))

40


In [15]:
print(shakespeare_nextwords['rose.'])

[]


**3. Make your own Shakespeare play**

Okay, that may be a bit of an exaggeration. Anyway, next up is text generation.
We'll need the choice function:

In [16]:
from random import choice

This function chooses a random element from a list.  Run this code a few times:

In [17]:
choice([1, 17, 34, 289])

289

Now, create a function called `makePlay(nextwords,n)` which takes a dictionary `nextwords` and a number `n`, and follows these steps:
1. Define an empty `newplay` string.  (I did this for you.)
2. Set `currentword` equal to '%'.
3. Start a for-loop that will run `n` times.
4. Randomly choose a word from the list `nextwords[currentword]`.
5. Add the chosen word to `newplay`, followed by a space `' '` character.
6. If the chosen word ends in a period, add a newline character '\n' and set `currentword` equal to `%`.
7. Otherwise, set `currentword` equal to the chosen word.
8. Go back to step 4.
9. When all `n` iterations have finished, your function should return `newplay`.

In [18]:
def makePlay(nextwords,n):
  newplay=''
  currentword = '%'
  for _ in range(n):
      if currentword not in nextwords or not nextwords[currentword]:
          currentword = '%'
          continue
      chosen_word = choice(nextwords[currentword])

      newplay += chosen_word + ' '

      if chosen_word.endswith('.'):
          newplay += '\n'
          currentword = '%'
      else:
          currentword = chosen_word

  return newplay


Now create a 1000 word Shakesperian play by running the following code:

In [19]:
nextwords=createNextwords(shakespeare_words)
newplay=makePlay(nextwords,1000)
print(newplay)

PROVOST. 
A SHIP BOATSWAIN MASTER. 
Pass our general MENELAUS, ULYSSES, DIOMEDES, TROILUS and made me. 
Come, the wenches may he loves, A merriment than noble man and earth with the watch. 
KING EDWARD. 
QUEEN. 
Rise, Titus, thou ow'st not; although, until we feel, to the stones new state by me; Behold and yield, Winchester; for you, a cloak? I be divided? Must I anatomize Among a golden sleep? VARRO. 
What I pray a stage, and show it; whereupon, after seven sons, what to th' event, then his birth. 
France. 
Ay, so fell a-doting, And is an oath, and when he comes. 
Romeo, good acts my foot, And youthful hose, a witty a mightier troops of nature's infinite tongue, who should Pour in her very fangs of deceits? ALICE. 
PHEBE. 
Then let him of Africa and that would say; I pray: sweet Hamlet! Lord bless thee more Upon these mutualities so plentifully gives it is stirring, worthy yet thy father. 
Where the rest I'd have made our supplications to be the King. 
The soil, the King miscarry. 
WI

At this point you might enjoy finding other sources of text on the internet and replacing the url at the top of this notebook. You can use song lyrics, news articles, etc. You just need to find files ending with '.txt'. Alternatively, you can copy and paste any block of text from any file into colab, assign it to the variable `text`, and then use wordlist=text.split() to get a list of words.

## Bonus:

Recall the following recursive function that we used to define the Fibonacci numbers:

In [20]:
def fibonacci(n):
  # Base cases: F(0) = 0, F(1) = 1
  if n == 0:
    return 0
  elif n == 1:
    return 1
  else:
    # Recursive case: F(n) = F(n-1) + F(n-2)
    return fibonacci(n - 1) + fibonacci(n - 2)

fibonacci(17)


1597

Here is the question.  If you run this line of code:  
`fibonacci(17)`  
how many times does the function `fibonacci()` get called?  
(The answer is not 1597, which is the value returned by the function.)

In [26]:
def fibonacci_calls(n):
    if n == 0:
        return 1
    elif n == 1:
        return 1
    else:
        return 1 + fibonacci_calls(n - 1) + fibonacci_calls(n - 2)

fibonacci_calls(17)

# The function gets called 5167 times. this happens because each call of the function
# leads to two more calls until it reaches the base cases, causing a rapid increase in
# total calls due to the recursive function.

5167

Reference:

[Counts](https://stackoverflow.com/questions/54714945/create-a-function-that-will-increment-by-one-when-called)